nx/source/services/capsa.c

```diff
@@ -0,0 +1,160 @@
+#define NX_SERVICE_ASSUME_NON_DOMAIN
+#include <string.h>
+#include <time.h>
+#include "service_guard.h"
+#include "runtime/hosversion.h"
+#include "services/applet.h"
+#include "services/capsa.h"
+
+static Service g_capsaSrv;
+
+NX_GENERATE_SERVICE_GUARD(capsa);
+
+Result _capsaInitialize(void) {
+    return smGetService(&g_capsaSrv, "caps:a");
+}
+
+void _capsaCleanup(void) {
+    serviceClose(&g_capsaSrv);
+}
+
+Service* capsaGetServiceSession(void) {
+    return &g_capsaSrv;
+}
+
+Result capsaGetAlbumFileCount(CapsAlbumStorage storage, u64* count) {
+    return serviceDispatchInOut(&g_capsaSrv, 0, storage, *count);
+}
+
+Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsApplicationAlbumEntry* buffer, u64 buffer_size) {
+    return serviceDispatchInOut(&g_capsaSrv, 1, storage, *count,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { buffer, buffer_size } },
+    );
+}
+
+Result capsaLoadAlbumFile(CapsAlbumFileId file_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size) {
+    return serviceDispatchInOut(&g_capsaSrv, 2, file_id, *out_size,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { jpeg_buffer, jpeg_buffer_size } },
+    );
+}
+
+Result capsaDeleteAlbumFile(CapsAlbumFileId file_id) {
+    return serviceDispatchIn(&g_capsaSrv, 3, file_id);
+}
+
+Result capsaStorageCopyAlbumFile(CapsAlbumFileId file_id, CapsAlbumStorage dst_storage) {
+    return serviceDispatchIn(&g_capsaSrv, 4, dst_storage);
+}
+
+Result capsaIsAlbumMounted(CapsAlbumStorage storage, bool* is_mounted) {
+    return serviceDispatchInOut(&g_capsaSrv, 5, storage, *is_mounted);
+}
+
+Result capsaGetAlbumUsage(CapsAlbumStorage storage, CapsAlbumUsage2 *out) {
+    return serviceDispatchInOut(&g_capsaSrv, 6, storage, *out);
+}
+
+Result capsaGetAlbumFileSize(CapsAlbumFileId file_id, u64* size) {
+    return serviceDispatchInOut(&g_capsaSrv, 7, file_id, *size);
+}
+
+Result capsaLoadAlbumFileThumbnail(CapsAlbumFileId file_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size) {
+    return serviceDispatchInOut(&g_capsaSrv, 8, file_id, *out_size,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { jpeg_buffer, jpeg_buffer_size } },
+    );
+}
+
+Result capsaGetAlbumUsage3(CapsAlbumStorage storage, CapsAlbumUsage3 *out) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    return serviceDispatchInOut(&g_capsaSrv, 15, storage, *out);
+}
+
+Result capsaGetAlbumMountResult(CapsAlbumStorage storage) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    return serviceDispatchIn(&g_capsaSrv, 17, storage);
```

Comment:
There's a bunch of dup funcs for this, use something like: `static Result _capsaCmdInU8NoOut(Service* srv, u8 inval, u32 cmd_id)`